### ISSUE TITLE
CoreBundle OrderRepository wrong constructor type-hint (EntityManager instead of EntityManagerInterface)

### ISSUE THREAD

**Issue by acornforth** [Sylius/Sylius#13549]
https://github.com/Sylius/Sylius/blob/ff02e5c88bc8df5b16434d4bc02568b47274ef3e/src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php#L36

Should use `EntityManagerInterface` instead.

This is preventing the use of EntityManagerDecorator to inject custom overridden entitymanagers

It shouldn't be a breaking change, and therefore potentially a fix could be rolled out for all recent enough versions of sylius

**EDIT:**

I can see that the current type-hint has been chosen because the constructor of  AssociationHydrator also is using the concrete type instead of the interface.

I'm using a approach like this in ProductRepository, PromotionRepository and OrderRepository to workaround it for now. 

```php
public function __construct(EntityManagerInterface $entityManager, Mapping\ClassMetadata $class)
{
	parent::__construct($entityManager, $class);

	if ($entityManager instanceof \App\Doctrine\CustomEntityManager) {
		$entityManager = $entityManager->getWrapped();
	}
	assert($entityManager instanceof \Doctrine\ORM\EntityManager);
	$this->associationHydrator = new AssociationHydrator($entityManager, $class);
}
```

But in order for this to work, I've had to make my App\Repository classes extend directly from the ones in ProductBundle, PromotionBundle and OrderBundle respectively, copying the CoreBundle methods into my App\Repository classes

**Comment by acornforth** [comment]
@Zales0123 I see your PR has been merged.

I'm happy to submit a PR for the fix here, but not sure which branch to base off, i guess it depends which Sylius version will pull in the changes to AssociationHydrator?

I anyone can advise me, i will get on the case :)

**Comment by Rafikooo** [comment]
This change will be available with the release of Sylius 2.0 💃
